hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/resolver/backend/AbstractBackendResolver.java

```diff
@@ -3,46 +3,117 @@
 import static org.hypertrace.traceenricher.util.EnricherUtil.createAttributeValue;
 
 import com.google.common.base.Joiner;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
+import org.hypertrace.core.datamodel.AttributeValue;
 import org.hypertrace.core.datamodel.Event;
 import org.hypertrace.core.datamodel.shared.HexUtils;
 import org.hypertrace.core.datamodel.shared.StructuredTraceGraph;
+import org.hypertrace.core.datamodel.shared.trace.AttributeValueCreator;
 import org.hypertrace.entity.constants.v1.BackendAttribute;
 import org.hypertrace.entity.data.service.v1.Entity.Builder;
 import org.hypertrace.entity.service.constants.EntityConstants;
 import org.hypertrace.entity.v1.entitytype.EntityType;
 import org.hypertrace.traceenricher.enrichedspan.constants.EnrichedSpanConstants;
 import org.hypertrace.traceenricher.enrichedspan.constants.v1.Backend;
 import org.hypertrace.traceenricher.enrichment.enrichers.BackendType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** Abstract class with common methods for all the different types of backend resolvers */
 public abstract class AbstractBackendResolver {
```

Comment:
Would it make sense to get rid of inheritance in the resolvers?

Reply:
Which inheritance? 

I guess you mean the inheritance on `AbstractBackendResolver`. But, that was the whole motivation of having `AbstractBackendResolver`, so that the common backend `resolve` logic can lie at a single place, and the real backend resolvers can just pass in the appropriate values, like backend URI, backend type, backend operation, backend destination etc.

Reply:
Can't the common logic present in `AbstractBackendResolver` be kept in some utility class or in some other class which implement the required methods and then used over in the resolvers using class composition?

Reply:
The common logic of `AbstractBackendResolver` can be moved to `AbstractBackendEntityEnricher` as well, and we can just convert `AbstractBackendResolver` into an interface, which the backend resolvers would implement

Did you mean that?

Reply:
> we can just convert AbstractBackendResolver into an interface, 

yes I meant this, but the logic in `AbstractBackendResolver` shouldn't be moved into `AbstractBackendEntityEnricher` instead that logic can be put into another class say `ResolverUtil` 
Basically the idea was to prefer class composition over inheritance (https://en.wikipedia.org/wiki/Composition_over_inheritance).

Reply:
And, who calls `ResolverUtil`?

Reply:
So all the resolvers use this Util class inside their implementation of `resolve` method

Reply:
And, that would be a common duplicate code across all resolvers?

```
public Optional<BackendInfo> resolve(Event event, StructuredTraceGraph structuredTraceGraph) {
  return ResolverUtil.resolve(event, structuredTraceGraph);   
}
```

Reply:
I am not sure, if this is even an example of composition over inheritance. Usually, composition over inheritance is preferred if you want to inherit multiple classes, which Java doesn't support, or the implementations of interface uses the super class, basically calling super in their methods

Reply:
> I am not sure, if this is even a fair example of composition over inheritance. 

Yeah I agree, not the perfect case.

> That being said, I can get rid of AbstractBackendResolver, and move the resolve logic inside AbstractBackendEntityEnricher

I feel, it's better to keep `resolve` method at the Resolver layer.

Reply:
> I feel, it's better to keep resolve method at the Resolver layer.

I think we are confusing, because of the naming. The backend resolvers are no longer the resolvers. They are just backend providers. And, the logic to do whatever you want to do with the information provided from those providers stays at a common place i.e. BackendEnricher

Basically, a `BackendEnricher` uses a set of backend providers, which provides information to `BackendEnricher` and `BackendEnricher` enriches that information in entity, event and trace

Providers should never (_according to me_) enrich the information in entity or event (_which it used to do before_). That is the responsibility of the backend enricher

Reply:
> Providers should never (according to me) enrich the information in entity or event (which it used to do before). That is the responsibility of the backend enricher

Yeah definitely 

> The backend resolvers are no longer the resolvers. They are just backend providers. 

Yeah, I got that from the implementation, basically resolution is defined as inference of backend from a set of attributes. 
Say you were to move this `resolve` method to `AbstractBackendEntityEnricher` method, do the individual resolver extend this class now?

Reply:
> Yeah, I got that from the implementation, basically resolution is defined as inference of backend from a set of attributes.

Right.

> Say you were to move this resolve method to AbstractBackendEntityEnricher method, do the individual resolver extend this class now?

The individual resolver (_would prefer to rename it to provider_) will not extend `AbstractBackendResolver` at all. In fact, there won't exist any abstract class `AbstractBackendResolver`. 
The individual resolver (_provider_) would just implement an interface, and that would just be `BackendProvider` interface